crates/fefix/src/session/event_loop.rs

```diff
@@ -81,13 +86,15 @@ where
                         }
                         Ok(num_bytes) => {
                             buf_filled_len += num_bytes;
-                            if buf_filled_len < buf.len() {
+                            self.decoder.add_bytes_read(num_bytes);
```

Comment:
This is the big change in the decoder - it is quite annoying, but couldn't think of how to do it better.

The issue is that the previous code relied on the buffer to be fully filled on each read (this works great with `read_exact` and in tests). The problem is that in the loop you cannot use `read_exact` because you can be interrupted by other async tasks like timers before you actually have read the whole amount.

Roughly what happened and caused the code to fail when running in real life is the following:
- on first iteration: decoder.fillable() returns an initial size of 20 bytes to read the header
- let's say 10 bytes get read before the heartbeat time interrupts the loop and causes the `next_event` function to exit.
- on second iteration: code will assume the whole 20 bytes were filled and continues appending

This results in having a 10 byte gap.

There needs to be a decoupling between the len of the buffer slice you read into and how many bytes have actually been read. Because of the decoupled manner of simply reading bytes into the decoder buffer, without the decoder actually being aware, the only way I could think of keeping track of how many bytes have actually been read is by adding this `add_bytes_read` method on the decoder.